### ISSUE TITLE
[SUGGESTION] Reduce InputHandler's CPU Usage

### ISSUE THREAD

**Issue by tdeeb** [teamradish/TRTwitchPlaysBot#49]
**Describe your suggestion**
On both GNU/Linux and Windows, running any input through the `InputHandler` causes a remarkable spike in CPU usage.

The thread taken from the thread pool is not throttled at all unlike the main application. This is to ensure a high degree of accuracy when processing inputs.

Adding `Thread.Sleep(1)` above [this line](https://github.com/teamradish/TRTwitchPlaysBot/blob/b738d1e3fb98832e7c0a422d553652f5fb014e48/TRBot/TRBot.Misc/InputHandler.cs#L268) fixes the CPU usage, but the accuracy goes down slightly. On my machine with not much else running, it was 1 ms off from my testing, but it's bound to stray more on less powerful machines or with more running, such as streaming through OBS while running a game.

I have an idea for improving this:
1. Sort input subsequences by duration, with lower durations being first. Do this while creating the array before starting the new thread for the input handler.
2. In the loop, instead of constantly checking for the stopwatch's time, instead sleep for the duration of the stopwatch's current timer subtracted by the input's duration. In simple terms, `Thread.Sleep((int)sw.ElapsedMilliseconds - input.duration)`. If the result of the subtraction is 0 or lower, move onto releasing the input since there's no reason to sleep.

With this, the inputs should be released at the correct times without using up all the CPU.

**HOWEVER**
Now we lose the ability to stop all running inputs! While the thread is asleep, it will no longer be checking for `InputsHalted` and exiting the loop. Furthermore, the order the inputs are pressed in subsequences is no longer deterministic and now depends on their duration; this is not ideal.

Our only potential option to have both low CPU usage and the ability to stop running inputs is to refactor `InputHandler` to utilize Tasks with CancellationTokens, and instead of using `Thread.Sleep`, use `Task.Delay` with the same CancellationToken. Look into this and see how well it works out.

**Comment by tdeeb** [comment]
This has been completed in https://github.com/teamradish/TRTwitchPlaysBot/commit/e6b1aefd337873efb0e017589a8c0ef1271c6651.

There's still some room for improvement (Ex. sleeping for longer based on the minimum time remaining), but the implementation there has noticeably lowered the CPU usage on my machine while retaining the same accuracy as before. It'll still spike up if there are lots of inputs going on (~5 or more every second), but it'll ramp up less than before.

I tried using `Task`, but there were issues canceling all of them with `Task.WhenAll`, namely it would throw a `TaskCanceledException` irregularly. On top of that, the accuracy was very far off, being **~2 to ~500+ milliseconds off sometimes when spamming long inputs!** `WaitAll` would stop the entire thread, which is not what I was looking for.

In short, https://github.com/teamradish/TRTwitchPlaysBot/commit/e6b1aefd337873efb0e017589a8c0ef1271c6651 implements the following:
1. In the loop, get the remaining time to press the input: `input.duration = sw.ElapsedMilliseconds`.
2. If this value is greater than 0, we're not pressing the input, so continue. However, if the value is less than our CPU save value, currently 20 milliseconds, then mention that we **should not** save CPU time.
3. If all inputs in the subsequence were greater than 20 milliseconds, then sleep the thread for 1 millisecond.

This ensures that we sleep when there's plenty of time remaining on the inputs and don't when it's approaching time to release one.
